connect/file/src/main/java/org/apache/kafka/connect/file/FileStreamSourceConnector.java

```diff
@@ -101,4 +105,50 @@ public ExactlyOnceSupport exactlyOnceSupport(Map<String, String> props) {
                 : ExactlyOnceSupport.UNSUPPORTED;
     }
 
+    @Override
+    public boolean alterOffsets(Map<String, String> connectorConfig, Map<Map<String, ?>, Map<String, ?>> offsets) {
+        AbstractConfig config = new AbstractConfig(CONFIG_DEF, connectorConfig);
+        String filename = config.getString(FILE_CONFIG);
+        if (filename == null || filename.isEmpty()) {
+            // If the 'file' configuration is unspecified, stdin is used and no offsets are tracked
+            throw new ConnectException("Offsets cannot be modified if the '" + FILE_CONFIG + "' configuration is unspecified. " +
+                    "This is because stdin is used for input and offsets are not tracked.");
+        }
+
+        // This connector makes use of a single source partition at a time which represents the file that it is configured to read from.
+        // However, there could also be source partitions from previous configurations of the connector.
+        for (Map.Entry<Map<String, ?>, Map<String, ?>> partitionOffset : offsets.entrySet()) {
+            Map<String, ?> partition = partitionOffset.getKey();
+            if (partition == null) {
+                throw new ConnectException("Partition objects cannot be null");
+            }
+
+            if (!partition.containsKey(FILENAME_FIELD)) {
+                throw new ConnectException("Partition objects should contain the key '" + FILENAME_FIELD + "'");
+            }
+
+            Map<String, ?> offset = partitionOffset.getValue();
+            // null offsets are allowed and represent a deletion of offsets for a partition
+            if (offset == null) {
+                return true;
+            }
+
+            if (!offset.containsKey(POSITION_FIELD)) {
+                throw new ConnectException("Offset objects should either be null or contain the key '" + POSITION_FIELD + "'");
+            }
+
+            // The 'position' in the offset represents the position in the file's byte stream and should be a non-negative long value
+            try {
+                long offsetPosition = Long.parseLong(String.valueOf(offset.get(POSITION_FIELD)));
```

Comment:
Ah, good catch, thanks. I think it might probably be a bit friendlier if we update the task class instead to do similar parsing, WDYT? I'm okay either way, since the most common use case would be copy pasting the output from `GET /offsets` and modifying it in which case users would end up using a number rather than a string anyway.

Reply:
I'd prefer to leave the task parsing the same; less work on our part, and less risk of a regression in existing parts of the code base.

Reply:
I found something interesting when experimenting with this. When passing a request body to the `PATCH /offsets` endpoint like:

```
{
  "offsets": [
    {
      "partition": {
        "filename": "/path/to/filename"
      },
      "offset": {
        "position": 20
      }
    }
  ]
}
```

the position `20` is deserialized to an `Integer` by Jackson (the JSON library we use in the REST layer for Connect) which seems fine because JSON doesn't have separate types for 32 bit and 64 bit numbers. So, the offsets map that is passed to `FileStreamSourceConnector::alterOffsets` by the runtime also contains `20` as an `Integer` value. I initially thought that this would cause the `FileStreamSourceTask` to fail at startup because it uses an `instanceof Long` check [here](https://github.com/apache/kafka/blob/aafbe3444354cfb0e4a8fdbd3443a63ba867c732/connect/file/src/main/java/org/apache/kafka/connect/file/FileStreamSourceTask.java#L95C58-L95C58) (and an `Integer` value is obviously not an instance of `Long`). However, interestingly, the task did not fail and doing some debugging revealed that after the offsets are serialized and deserialized by the `JsonConverter` in `OffsetsStorageWriter` (in `Worker::modifySourceConnectorOffsets`) and `OffsetsStorageReader` respectively, the offsets map that is retrieved by the task on startup through its context contains the position `20` as a `Long` value. 

While this particular case is easily handled by simply accepting `Integer` values as valid in the `FileStreamSourceConnector::alterOffsets` method, I'm thinking we probably need to make some changes so that the offsets map passed to source connectors in their `alterOffsets` method is the same as the offsets map that connectors / tasks will retrieve via the `OffsetsStorageReader` from their context (otherwise, this could lead to some hard to debug issues in other connectors implementing the `SourceConnector::alterOffsets` method). The easiest way off the top of my head would probably would be to serialize and deserialize the offsets map using the `JsonConverter` before invoking `SourceConnector::alterOffsets`. WDYT?

Furthermore, just checking whether the offset position is an instance of `Long` (Jackson uses a `Long` if the number doesn't fit in an `Integer`) or `Integer` in the `FileStreamSourceConnector::alterOffsets` method seems sub-optimal because:

- To someone just reading through the `FileStreamSourceConnector` and `FileStreamSourceTask` classes, accepting `Integer` instances during validation but requiring `Long` instances in the actual task would look like a bug since the serialization + deserialization aspect isn't transparent.
- It's an extremely unlikely scenario, but any changes in Jackson's deser logic could break things here - for instance, if smaller numbers are deserialized into `Short`s instead of `Integer`s. Parsing using a combination of `String::valueOf` and `Long::parseLong` seems a lot more robust.

Reply:
Ah, nice catch! I noticed the discrepancy in numeric types while working on [KAFKA-15177](https://issues.apache.org/jira/browse/KAFKA-15177) but hadn't even considered the possibility of aligning the types across invocations of `alterOffsets` and `OffsetStorageReader::offset`/`OffsetStorageReader::offsets`.

I think re-deserializing the offsets before passing them to `alterOffsets` is a great idea. Unless the request body is gigantic there shouldn't be serious performance concerns, and it also acts as a nice preflight check to ensure that the offsets can be successfully propagated to the connector's tasks through the offsets topic.

I still don't love permitting string types for the connector's `position` offset values--it doesn't seem like a great endorsement of our API if we have to implement workarounds in the file connectors, which are the first example of the connector API that many developers see.

Reply:
Since the type alignment issue seemed like a broader one (i.e. not scoped to the file connectors being touched here), I've created a separate [ticket](https://issues.apache.org/jira/browse/KAFKA-15182) and [PR](https://github.com/apache/kafka/pull/14003) for it.

> it doesn't seem like a great endorsement of our API if we have to implement workarounds in the file connectors, which are the first example of the connector API that many developers see.

I'd argue that it isn't really a workaround and that the current check itself is bad. If the (de)serialization happened to use `Integer` for values that fit in a 32 bit signed type (which would be perfectly valid and is exactly what we see currently before the values are passed through the converter), the current check in the `FileStreamSourceTask` would cause it to bail.

Reply:
Hmmm... wouldn't that be a pretty serious breaking change if we accidentally switched up how the JSON converter deserializes integer types? Not just for the file source connector, but for plenty of others.

It feels like it might be a better use of our time to make note of this possibility and ensure that we have sufficient unit testing in place to prevent that kind of regression (I suspect we already do but haven't verified this yet).

Reply:
> Hmmm... wouldn't that be a pretty serious breaking change if we accidentally switched up how the JSON converter deserializes integer types? Not just for the file source connector, but for plenty of others.

Okay, that's fair enough, I've changed the check in `FileStreamSourceConnector::alterOffsets` to mirror the one made in the task at startup for consistency (and avoided making changes in the existing task logic).

> I don't know if this significantly changes the conversation but it seems subtle and counterintuitive enough to bring up so that we can avoid accidentally breaking connector code that relies on this behavior.

Hm yeah, that's definitely another interesting one to bring up - however, I'd contend that that one kinda makes sense since we're passing the `SourceRecord` itself - tasks already deal with `SourceRecord` and their offsets (and associated types) in their regular lifecycle. It would be highly confusing if the `SourceRecord` that they get in `commitRecord` doesn't match the one they dispatched to the framework via `poll`. Of course, ideally, the offsets that they read via `OffsetStorageReader` should also not have any type mismatches compared to the `SourceRecord` ones, but I don't think we'd want to (or safely could) change that at this point.

Since the offsets being altered externally would correspond to the ones that the connector / tasks read at startup, I think it makes sense to align the types across invocations to `SourceConnector::alterOffsets` and offsets queried from an `OffsetStorageReader` (and an implicit separate alignment between the `SourceRecord`'s offsets types).

Reply:
💯 Sounds good, thanks for thinking this through.